### PR TITLE
feat(gateway): initial support for GatewayStatus.Listeners

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

This PR covers the `ResolvedRefs` (i.e. checking certificate references for `ReferencePolicy`) and `Ready` conditions.

Part of https://github.com/kumahq/kuma/issues/3617

### CODE REVIEW

**Review by jpeach — CHANGES_REQUESTED** [review]
FWIW it would be helpful to include more detail about the change in the commit message (mirrored in the PR) ... it wasn't clear to me that this PR addresses refpolicy for secrets until I went back and looked at the diff for each commit. It was very hard to tell from the aggregated diff because of the code movement.

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+		listener.Hostname = "*"
+		if l.Hostname != nil {
+			listener.Hostname = string(*l.Hostname)
+		}
+
+		refsPermitted := true
+
+		for _, certRef := range l.TLS.CertificateRefs {
+			ref := policy.PolicyReferenceSecret(policy.FromGatewayIn(gateway.Namespace), *certRef)
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:58]: Yeah, that's the result of changing the API and code too many times. Fixed the NPE in a separate PR.

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+					{
+						Type:    string(gatewayapi.ListenerConditionReady),
+						Status:  kube_meta.ConditionFalse,
+						Reason:  string(gatewayapi.ListenerReasonInvalid),
+						Message: "unable to resolve refs",
+					},
+				}
+			}
+
+			listenerConditions[l.Name] = append(listenerConditions[l.Name], resolvedCondition...)
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:100]: Of course :man_facepalming:

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+					},
+				}
+			} else {
+				targetGK := kube_schema.GroupKind{Group: string(*certRef.Group), Kind: string(*certRef.Kind)}
+				resolvedCondition = []kube_meta.Condition{
+					{
+						Type:    string(gatewayapi.ListenerConditionResolvedRefs),
+						Status:  kube_meta.ConditionFalse,
+						Reason:  string(gatewayapi.ListenerReasonRefNotPermitted),
+						Message: fmt.Sprintf("reference to %s %s/%s not permitted by any ReferencePolicy", targetGK.String(), ref.ToNamespace(), certRef.Name),
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:89]: > nit: if you take the name and namespace from the same object it is more obvious

the issue is that `certRef.Namespace` may be nil, that's why I introduced `PolicyReference`. I'll expose the other fields there.

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/policy/policy.go [hunk 4]
```diff
@@ -5,12 +5,12 @@ import (
...
 
 	"github.com/pkg/errors"
-	kube_schema "k8s.io/apimachinery/pkg/runtime/schema"
 	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
 	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
 )
 
-var httpRouteGK = kube_schema.GroupKind{Group: gatewayapi.GroupName, Kind: "HTTPRoute"}
+var httpRouteKind = gatewayapi.Kind("HTTPRoute")
+var gatewayKind = gatewayapi.Kind("Gateway")
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/policy/policy.go:13]: I think I prefer to leave them out. This makes it feel less like I'm referring to types with strings at least, and makes it more likely they'll be reused.

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	kube_meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_schema "k8s.io/apimachinery/pkg/runtime/schema"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/controllers/gatewayapi/policy"
+)
+
+type ListenerConditions map[gatewayapi.SectionName][]kube_meta.Condition
```
[the remark this answers is not in the document]
Comment (jpeach) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:16]: https://github.com/kubernetes-sigs/gateway-api/issues/984

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/policy/policy.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/policy/policy.go:13]: Actually this package is so small, you're right

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:89]: done, also improved the error code

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:58]: also improved the caller code

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:16]: Added to https://github.com/kumahq/kuma/issues/3617

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_status.go [hunk 6]
```diff
@@ -42,14 +46,53 @@ func mergeGatewayStatus(gateway *gatewayapi.Gateway, instance *mesh_k8s.GatewayI
...
+
+	gateway.Status.Listeners = mergeGatewayListenerStatuses(gateway, listenerConditions)
 
 	readinessStatus := kube_meta.ConditionFalse
 	readinessReason := gatewayapi.GatewayReasonListenersNotReady
 
 	// TODO(michaelbeaumont) it'd be nice to get more up to date info from the
 	// kuma-dp instance to tell whether listeners are _really_ ready
-	if len(addrs) == 0 {
+	if len(gateway.Status.Addresses) == 0 {
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_status.go:100]: I've got a PR ready for after this is merged

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go]: https://github.com/kumahq/kuma/pull/3693

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_status.go [hunk 7]
```diff
@@ -27,8 +32,7 @@ func (r *GatewayReconciler) updateStatus(ctx context.Context, gateway *gatewayap
 	return nil
 }
 
-// mergeGatewayStatus updates the status by mutating the given Gateway.
-func mergeGatewayStatus(gateway *gatewayapi.Gateway, instance *mesh_k8s.GatewayInstance) {
+func gatewayAddresses(gateway *gatewayapi.Gateway, instance *mesh_k8s.GatewayInstance) []gatewayapi.GatewayAddress {
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_status.go:35]: See below, doing in separate PR.

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpeach) [reply pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_conversion.go:16]: I was totally wrong about this - the section name is guaranteed to be unique and there are API server validations to enforce that.

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_conversion.go [hunk 8]
```diff
@@ -181,16 +171,15 @@ func (r *HTTPRouteReconciler) gapiToKumaRef(
...
-		if err := r.Client.Get(ctx, key, svc); err != nil {
+		if err := r.Client.Get(ctx, namespacedName, svc); err != nil {
 			if kube_apierrs.IsNotFound(err) {
 				return nil,
 					&kube_meta.Condition{
 						Type:    string(gatewayapi.ConditionRouteResolvedRefs),
 						Status:  kube_meta.ConditionFalse,
 						Reason:  ObjectNotFound,
-						Message: fmt.Sprintf("backend reference references a non-existent Service %s", key.String()),
+						Message: fmt.Sprintf("backend reference references a non-existent Service %s", namespacedName.String()),
```
Comment (jpeach) [inline pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_conversion.go:182]: Nit: prefer %q for unambiguously formatting names

File: pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_conversion.go [hunk 9]
```diff
@@ -138,34 +136,26 @@ func k8sToKumaHeader(header gatewayapi.HTTPHeader) *mesh_proto.GatewayRoute_Http
...
 		return nil, nil, errors.Wrap(err, "couldn't determine if backend reference is permitted")
 	} else if !permitted {
-		targetGK := kube_schema.GroupKind{Group: string(*ref.Group), Kind: string(*ref.Kind)}
 		return nil,
 			&kube_meta.Condition{
 				Type:    string(gatewayapi.ConditionRouteResolvedRefs),
 				Status:  kube_meta.ConditionFalse,
 				Reason:  RefNotPermitted,
-				Message: fmt.Sprintf("reference to %s %s/%s not permitted by any ReferencePolicy", targetGK.String(), policyRef.ToNamespace(), ref.Name),
+				Message: fmt.Sprintf("reference to %s %s not permitted by any ReferencePolicy", gk, namespacedName),
```
Comment (jpeach) [inline pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_conversion.go:152]: Nit: I wanna suggest using %q for the NamespacedName, but it's borderline, so if you don't agree, don't worry about it :)

**Review by jpeach — APPROVED** [review]

**Pull request merged**